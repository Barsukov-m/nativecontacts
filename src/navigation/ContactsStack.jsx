import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { CustomHeader } from '../components/CustomHeader';
import { capitalizeFirstLetter } from '../utils/contacts';

// Screens
import { ContactsScreen } from '../screens/Contacts';
import { DetailsScreen } from '../screens/Details';

const Stack = createNativeStackNavigator();

const ContactsStack = () => {
  return (
    <Stack.Navigator initialRouteName="Home">
      <Stack.Screen
        name="Home"
        component={ContactsScreen}
        options={({ navigation, route }) =>
          CustomHeader({
            navigation,
            route,
            title: 'Native Contacts',
          })
        }
      />
      <Stack.Screen
        name="Details"
        component={DetailsScreen}
        options={({ navigation, route }) => {
          const { contact } = route.params;
          const firstName = capitalizeFirstLetter(contact.name.first);
          const lastName = capitalizeFirstLetter(contact.name.last);

          return CustomHeader({
            navigation,
            route,
            title: `${firstName} ${lastName}`,
          });
        }}
      />
    </Stack.Navigator>
  );
};

export default ContactsStack;
