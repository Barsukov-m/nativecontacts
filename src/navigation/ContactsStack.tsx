import { useCallback } from 'react';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { CustomHeader } from '../components/CustomHeader';
import { ContactInterface } from '../types/navigationTypes';
import { getHeaderTitleByContact } from '../utils/contactsUtils';

// Screens
import { ContactsScreen } from '../screens/Contacts';
import { DetailsScreen } from '../screens/Details';

const Stack = createNativeStackNavigator();

interface ContactsStackProps {
	navigation: {
		goBack?: () => void;
		openDrawer?: () => void;
	};
	route: {
		name: string;
		params?: {
			contact?: ContactInterface;
		};
	};
}

const ContactsStack = () => {
	const homeOptions = useCallback(
		({ navigation, route }: ContactsStackProps) => ({
			header: () => <CustomHeader title="Native Contacts" route={route} />,
		}),
		[]
	);

	const detailsOptions = useCallback(
		({ navigation, route }: ContactsStackProps) => ({
			header: () => {
				const { contact } = route.params as { contact: ContactInterface };
				const title = getHeaderTitleByContact(contact);

				return <CustomHeader title={title} route={route} />;
			},
		}),
		[getHeaderTitleByContact]
	);

	return (
		<Stack.Navigator initialRouteName="Home">
			<Stack.Screen
				name="Home"
				component={ContactsScreen}
				options={homeOptions}
			/>
			<Stack.Screen
				name="Details"
				component={DetailsScreen}
				options={detailsOptions}
			/>
		</Stack.Navigator>
	);
};

export default ContactsStack;
