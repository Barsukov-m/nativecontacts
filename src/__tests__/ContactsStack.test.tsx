import { NavigationContainer } from '@react-navigation/native';
import { fireEvent } from '@testing-library/react-native';
import ContactsStack from '../navigation/ContactsStack';
import {
	renderContactsOnScreen,
	renderWithProviders,
} from 'utils/reduxTestUtils';

describe('<ContactsStack />', () => {
	const renderComponent = () =>
		renderWithProviders(
			<NavigationContainer>
				<ContactsStack />
			</NavigationContainer>
		);

	it('renders correctly', () => {
		const tree = renderComponent();
		const { getByText } = tree;

		expect(getByText('Native Contacts')).toBeDefined();

		expect(tree).toMatchSnapshot();
	});

	it('takes you to the details screen when clicking on a contact', async () => {
		const { getAllByTestId, getByText } = renderComponent();
		await renderContactsOnScreen();

		const contactImages = getAllByTestId('contact-list-image');

		expect(contactImages).toBeDefined();

		expect(contactImages.length).toBeGreaterThan(0);

		fireEvent(contactImages[0], 'press');
		expect(getByText('Phone')).toBeDefined();
	});
});
