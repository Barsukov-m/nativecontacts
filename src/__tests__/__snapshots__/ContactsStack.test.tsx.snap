// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactsStack /> renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onGestureCancel={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={true}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="Home"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "height": "100%",
                "justifyContent": "center",
              }
            }
          >
            <ActivityIndicator
              size="large"
            />
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#FAFAFA",
                "elevation": 5,
                "flexDirection": "row",
                "height": 56,
                "justifyContent": "space-between",
                "paddingHorizontal": 16,
                "shadowColor": "#212121",
                "shadowOffset": {
                  "height": 2,
                  "width": 0,
                },
                "shadowOpacity": 0.25,
                "shadowRadius": 3.84,
              }
            }
          >
            <View
              style={
                {
                  "flexDirection": "row",
                }
              }
            >
              <Text>
                Icon
              </Text>
              <Text
                style={
                  {
                    "color": "#212121",
                    "fontSize": 20,
                    "fontWeight": "500",
                  }
                }
              >
                Native Contacts
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
